### PR TITLE
Np 1644/fix dot method

### PR DESCRIPTION
**Pull request by olhabn**
fixing NestedPropertiesExpression

@lchanmann @kgrgreer @jlhughes 
please review

### CODE REVIEW

**Comment by lchanmann** [comment]
@olhabn would be great if you could implement REF and use REF in the nested property expression instead of DOT. Please follow the discussion on dev-team channel for the distinction between DOT and REF.

**Comment by olhabn** [comment]
> @olhabn would be great if you could implement REF and use REF in the nested property expression instead of DOT. Please follow the discussion on dev-team channel for the distinction between DOT and REF.

will do. thanks!

File: src/foam/mlang/mlang.js [hunk 1]
```diff
@@ -3248,39 +3293,23 @@ foam.CLASS({
...
-          try {
-            return getArg2().f(obj1);
-          } catch ( Throwable t ) {
-            Logger logger = (Logger) getX().get("logger");
-            if ( logger == null ) {
-              logger = new StdoutLogger();
-            }
-            logger.error(t);
-            return null;
+        StringBuilder sb = new StringBuilder("find");
```
Comment (olhabn) [inline src/foam/mlang/mlang.js]: let me know if i should refactor it to use invokeFinderMethod as Chanmann suggested for Dot in https://nanopay.slack.com/archives/G0336TYKV/p1594412722054800

File: src/foam/mlang/mlang.js [hunk 2]
```diff
@@ -3225,6 +3225,49 @@ foam.CLASS({
...
+
+  methods: [
+    {
+      name: 'f',
+      code: function(o) {
+        return this.arg2.f(this.arg1.f(o));
+      },
+      javaCode: `
+        Object e1 = getArg1().f(obj);
+        if ( e1 == null ) return e1;
```
Comment (olhabn) [inline src/foam/mlang/mlang.js]: the reason for this change is that 'Ref' could return null

File: src/foam/nanos/column/ColumnConfigPropertyValue.js [hunk 3]
```diff
@@ -81,7 +81,8 @@ foam.CLASS({
               if ( p instanceof foam.core.AbstractFObjectPropertyInfo ) {
                 cls = p.getValueClass();
               } else {
```
Comment (olhabn) [inline src/foam/nanos/column/ColumnConfigPropertyValue.js:83]: have no idea why it looked like that. maybe I reverted something on my columnconfig branch or something

File: src/foam/util/StringUtil.java [hunk 4]
```diff
@@ -76,6 +76,6 @@ public static String capitalize(String s) {
     if ( SafetyUtil.isEmpty(s) ) return s;
     char[] chars = s.toCharArray();
     chars[0] = Character.toUpperCase(chars[0]);
-    return chars.toString();
```
Comment (olhabn) [inline src/foam/util/StringUtil.java:79]: same here

File: src/foam/mlang/mlang.js [hunk 5]
```diff
@@ -3225,6 +3225,49 @@ foam.CLASS({
...
+         Compare property values using arg2's property value comparator.
+         Used by GroupBy
+      **/
+      return this.arg2.comparePropertyValues(o1, o2);
+    }
+  ]
+});
+
+
+foam.CLASS({
```
Comment (jlhughes) [inline src/foam/mlang/mlang.js]: Can you move this into its own source file, please - we are starting to break up these large js files.

File: src/foam/mlang/mlang.js [hunk 6]
```diff
@@ -3248,39 +3293,22 @@ foam.CLASS({
...
-            Logger logger = (Logger) getX().get("logger");
-            if ( logger == null ) {
-              logger = new StdoutLogger();
-            }
-            logger.error(t);
-            return null;
+        PropertyInfo p1 = (PropertyInfo) getArg2();
+        FObject refObj = null;
+        try {
+          refObj = (FObject)getArg1().getObjClass().getMethod("find" + StringUtil.capitalize(p1.getName()), foam.core.X.class).invoke(obj, ((FObject)obj).getX());
```
Comment (jlhughes) [inline src/foam/mlang/mlang.js]: Can you check if Relationship has a helper method to do this, so we only have method name creation logic in one place, please. 
if it doesn't, please add.

Comment (olhabn) [reply src/foam/mlang/mlang.js]: yes, it does, but it's not really clear to me where we could retrieve a relationship for a reference property. is there some kind of relationship dao? thanks
is there Relationship class for backend?

File: src/foam/mlang/mlang.js (hunk 5 above)
Comment (olhabn) [reply src/foam/mlang/mlang.js]: fixed

File: src/foam/mlang/mlang.js (hunk 6 above)
Comment (olhabn) [reply src/foam/mlang/mlang.js]: my understanding of relationship is that we use them for java code generation. so I'm wondering if we maybe create something like AbstractReferencePropertyInfo and copy info from relationship to it?

Comment (jlhughes) [reply src/foam/mlang/mlang.js]: @olhabn ok, leave that for now and create another ticket to capture the support method - and ideas.

File: src/foam/mlang/expr/Ref.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.mlang.expr',
+  name: 'Ref',
+  extends: 'foam.mlang.AbstractExpr',
+  implements: [ 'foam.core.Serializable' ],
+
+  documentation: 'A Binary Predicate which returns reference property object',
```
Comment (lchanmann) [inline src/foam/mlang/expr/Ref.js]: please update the documentation.

File: src/foam/mlang/expr/Ref.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+  javaImports: [
+    'foam.core.AbstractFObjectPropertyInfo',
```
Comment (lchanmann) [inline src/foam/mlang/expr/Ref.js]: unused import

File: src/foam/mlang/expr/Ref.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    {
+      name: 'f',
+      code: async function(o) {
+        return o[arg1.name + '$find'];
+      },
+      javaCode: `
+        PropertyInfo p1 = (PropertyInfo) getArg1();
+        FObject refObj = null;
+        try {
+          refObj = (FObject)obj.getClass().getMethod("find" + StringUtil.capitalize(p1.getName()), foam.core.X.class).invoke(obj, ((FObject)obj).getX());
```
Comment (lchanmann) [inline src/foam/mlang/expr/Ref.js]: can you break this line into two? one for getMethod and another to invoke.

File: src/foam/mlang/expr/Ref.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        return refObj;
+      `
+    },
+
+    function comparePropertyValues(o1, o2) {
+      /**
+         Compare property values using arg2's property value comparator.
+         Used by GroupBy
+      **/
+      return this.arg2.comparePropertyValues(o1, o2);
```
Comment (lchanmann) [inline src/foam/mlang/expr/Ref.js]: there is no arg2

File: src/foam/mlang/expr/Ref.js (hunk 7 above)
Comment (olhabn) [reply src/foam/mlang/expr/Ref.js]: fixed. thanks

File: src/foam/mlang/expr/Ref.js (hunk 8 above)
Comment (olhabn) [reply src/foam/mlang/expr/Ref.js]: fixed.

File: src/foam/mlang/expr/Ref.js (hunk 9 above)
Comment (olhabn) [reply src/foam/mlang/expr/Ref.js]: fixed.

File: src/foam/mlang/expr/Ref.js (hunk 10 above)
Comment (olhabn) [reply src/foam/mlang/expr/Ref.js]: fixed.

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 11]
```diff
@@ -81,61 +81,95 @@ foam.CLASS({
...
-        }
+          name: 'prop',
+          javaType: 'foam.core.PropertyInfo',
+          javaInfoType: 'foam.core.AbstractObjectPropertyInfo'
+        },
       ],
-      code: function (prop1, prop2) {
-        return foam.mlang.Expressions.create().DOT(prop1, prop2);
+      code: function (of, propName, i, expr) {
+        return false;
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: is it always false for js?

Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: it was kinda not implemented.  sorry about that I've cleaned it up.

File: src/foam/mlang/expr/Ref.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    {
+      class: 'foam.mlang.ExprProperty',
+      name: 'arg1'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'f',
+      code: async function(o) {
```
Comment (kgrgreer) [inline src/foam/mlang/expr/Ref.js]: JS predicates can't be async. Better to throw an exception.

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 13]
```diff
@@ -81,61 +81,92 @@ foam.CLASS({
...
-      code: function (of, propName, i) {
-        if ( i === propName.length - 1 )
-          return of.getAxiomByName(propName[i]);
+      code: function (of, propName, i, expr) {
         var prop = of.getAxiomByName(propName[i]);
-        return foam.mlang.Expressions.create().DOT(prop, this.returnPropExpr(prop, this.returnDotExprForNestedProperty(prop.of, propName, ++i)));
+        if ( ! prop ) return null;
+
+        var thisPropExpr; 
+        if ( foam.core.Reference.isInstance(prop) ) thisPropExpr = foam.mlang.Expressions.create().REF(prop);
```
Comment (kgrgreer) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: Please format this if/else and the next one normally (for FOAM). Thx.

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 14]
```diff
@@ -81,61 +81,92 @@ foam.CLASS({
...
+        if ( prop == null ) return null; 
+
+        Boolean isPropAReference = isPropertyAReference(ci, prop);
+
+        Expr thisPropExpr = null;
+        if ( isPropAReference )
+          thisPropExpr = REF(prop);
+        else
+          thisPropExpr = prop;
+        if ( expr == null) expr = thisPropExpr;
```
Comment (kgrgreer) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: same here.

File: src/foam/mlang/expr/Ref.js (hunk 12 above)
Comment (olhabn) [reply src/foam/mlang/expr/Ref.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 13 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 14 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

**Comment by lchanmann** [comment]
You can extract `buildPropertyExpr` method, something like
```
function buildPropertyExpr(prop, expr) {
  if ( foam.core.Reference.isInstance(prop) ) {
    prop = foam.mlang.Expressions.create().REF(prop);
  }

  return expr == null ? prop :
    foam.mlang.Expressions.create().DOT(expr, thisPropExpr);
}
```
and `returnDotExprForNestedProperty` would be shorter/easier to comprehend
```
function returnDotExprForNestedProperty(of, propName, i, expr) {
  var prop = of.getAxiomByName(propName[i]);
  if ( ! prop ) return null;

  var propExpr = buildPropertyExpr(prop, expr);

  if ( i === propName.length - 1 )
    return propExpr;

  return this.returnDotExprForNestedProperty(prop.of, propName, ++i, propExpr);
}
```
Same goes for Java and also extract `getPropertyClassInfo`.

File: src/foam/mlang/mlang.js (hunk 2 above)
Comment (lchanmann) [reply src/foam/mlang/mlang.js]: rename e1 to receiver and return null instead of the variable.

File: src/foam/mlang/expr/Ref.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.mlang.expr',
+  name: 'Ref',
+  extends: 'foam.mlang.AbstractExpr',
+  implements: [ 'foam.core.Serializable' ],
+
+  documentation: 'An Unary Predicate which returns reference property object',
```
Comment (lchanmann) [inline src/foam/mlang/expr/Ref.js]: REF is an expression not predicate.

Comment (olhabn) [reply src/foam/mlang/expr/Ref.js]: noted and fixed. thanks

**Comment by olhabn** [comment]
> You can extract `buildPropertyExpr` method, something like
> 
> ```
> function buildPropertyExpr(prop, expr) {
>   if ( foam.core.Reference.isInstance(prop) ) {
>     prop = foam.mlang.Expressions.create().REF(prop);
>   }
> 
>   return expr == null ? prop :
>     foam.mlang.Expressions.create().DOT(expr, thisPropExpr);
> }
> ```
> 
> and `returnDotExprForNestedProperty` would be shorter/easier to comprehend
> 
> ```
> function returnDotExprForNestedProperty(of, propName, i, expr) {
>   var prop = of.getAxiomByName(propName[i]);
>   if ( ! prop ) return null;
> 
>   var propExpr = buildPropertyExpr(prop, expr);
> 
>   if ( i === propName.length - 1 )
>     return propExpr;
> 
>   return this.returnDotExprForNestedProperty(prop.of, propName, ++i, propExpr);
> }
> ```
> 
> Same goes for Java and also extract `getPropertyClassInfo`.

fixed. please review

File: src/foam/mlang/mlang.js (hunk 2 above)
Comment (olhabn) [reply src/foam/mlang/mlang.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 16]
```diff
@@ -81,61 +81,118 @@ foam.CLASS({
...
       ],
-      code: function (of, propName, i) {
-        if ( i === propName.length - 1 )
-          return of.getAxiomByName(propName[i]);
+      code: function (of, propName, i, expr) {
         var prop = of.getAxiomByName(propName[i]);
-        return foam.mlang.Expressions.create().DOT(prop, this.returnPropExpr(prop, this.returnDotExprForNestedProperty(prop.of, propName, ++i)));
+        if ( ! prop ) return null;
+
+        expr = this.buildPropertyExpr(prop, expr);
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: Please create a new variable
```suggestion
        var propExpr = this.buildPropertyExpr(prop, expr);
```

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 17]
```diff
@@ -81,61 +81,118 @@ foam.CLASS({
...
-        PropertyInfo p = (PropertyInfo) ci.getAxiomByName(propName[i]);
-        if ( i == propName.length - 1 ) {
-          return p;
-        }
+        PropertyInfo prop = (PropertyInfo) ci.getAxiomByName(propName[i]);
+
+        if ( prop == null ) return null;
+
+        Boolean isPropAReference = isPropertyAReference(ci, prop);
+        expr = buildPropertyExpr(prop, expr, isPropAReference);
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: buildPropertyExpr should take a property and expression same as the JS one and in it you check if the property is a reference property.

I think you might not need classInfo argument for isPropertyAReference method.

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 18]
```diff
@@ -81,61 +81,118 @@ foam.CLASS({
...
+
+        if ( i == propName.length - 1 )
+          return expr;
         
-        StringBuilder sb = new StringBuilder("find");
         try {
+          StringBuilder sb = new StringBuilder("find");
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: can you extract the logic in try to method as well?

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 19]
```diff
@@ -81,61 +81,118 @@ foam.CLASS({
...
+        if ( isPropertyAReference )
+          thisPropExpr = REF(prop);
+        else
+          thisPropExpr = prop;
+        if ( expr == null )
+          expr = thisPropExpr;
+        else
+          expr = DOT(expr, thisPropExpr);
+
+        return expr;
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]:
```suggestion
        if ( isPropAReference(prop) ) {
          prop = REF(prop);
        }
        
        return expr == null ? prop : DOT(expr, prop);
```

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 20]
```diff
@@ -81,61 +81,118 @@ foam.CLASS({
...
         }
       ],
-      code: function (prop1, prop2) {
-        return foam.mlang.Expressions.create().DOT(prop1, prop2);
+      code: function(prop, expr) {
+        if ( foam.core.Reference.isInstance(prop) ) {
+          prop = foam.mlang.Expressions.create().REF(prop);
+        }
+      
+        return expr == null ? prop :
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: expr is `undefined` when not given
```suggestion
        return ! expr ? prop :
```

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 16 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed. thanks

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 17 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 18 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 19 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: have problem with casting it back to Property as of now

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 20 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 19 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 21]
```diff
@@ -81,61 +81,120 @@ foam.CLASS({
...
         ClassInfo ci = of;
-        PropertyInfo p = (PropertyInfo) ci.getAxiomByName(propName[i]);
-        if ( i == propName.length - 1 ) {
-          return p;
-        }
+        PropertyInfo prop = (PropertyInfo) ci.getAxiomByName(propName[i]);
+
+        if ( prop == null ) return null;
+
+        Boolean isPropAReference = isPropertyAReference(prop);
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: unused variable

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 22]
```diff
@@ -81,61 +81,120 @@ foam.CLASS({
...
+      `
+    },
+    {
+      name: 'isPropertyAReference',
+      javaType: 'Boolean',
+      args: [
+        {
+          name: 'prop',
+          javaType: 'foam.core.PropertyInfo',
+          javaInfoType: 'foam.core.AbstractObjectPropertyInfo'
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: javaInfoType not needed

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 23]
```diff
@@ -81,61 +81,120 @@ foam.CLASS({
...
+        if ( ci == null )
+          return null; 
         
-        StringBuilder sb = new StringBuilder("find");
+        return returnDotExprForNestedProperty(ci, propName, ++i, expr);
+      `
+    },
+    {
+      name: 'getPropertyValueClass',
+      javaType: 'ClassInfo',
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: it returns a class info, should it be getPropertyClassInfo?

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 24]
```diff
@@ -81,61 +81,120 @@ foam.CLASS({
...
           Class cls;
-          if ( p instanceof foam.core.AbstractFObjectPropertyInfo ) {
-            cls = p.getValueClass();
-          } else {
-            sb.setLength(4);
-            Method m = ci.getObjClass().getMethod(StringUtil.capitalize(p.getName()), foam.core.X.class);
+          if ( ! isPropAReference )
+            cls = prop.getValueClass();
+          else {
+            Method m = ci.getObjClass().getMethod("find" + StringUtil.capitalize(prop.getName()), foam.core.X.class);
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: if prop.getClassInfo() and ci are the same we can spare taking ci as argument.

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 25]
```diff
@@ -81,61 +81,120 @@ foam.CLASS({
...
+          javaType: 'foam.core.ClassInfo'
+        },
+        {
+          name: 'prop',
+          javaType: 'foam.core.PropertyInfo',
+        }
+      ],
+      javaCode: `
         try {
+          Boolean isPropAReference = isPropertyAReference(prop);
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: isPropertyAReference does a reflection call and this method does the same reflection call again on line 141. If we have something like `getFinderMethod(prop)` then reuse the result there would be one reflection call.

File: src/foam/mlang/mlang.js [hunk 26]
```diff
@@ -3223,16 +3223,7 @@ foam.CLASS({
...
-
-  javaImports: [
-    'foam.core.AbstractFObjectPropertyInfo',
-    'foam.core.FObject',
-    'foam.core.PropertyInfo',
-    'foam.nanos.logger.Logger',
-    'foam.nanos.logger.StdoutLogger',
-    'foam.util.StringUtil'
-  ],
+  documentation: 'A Binary Expression which applies arg2.f() to arg1.f().',
```
Comment (lchanmann) [inline src/foam/mlang/mlang.js]: while you are here, can you also change this to
```suggestion
  documentation: `
    A Binary Expression which evaluates arg1 and passes the result to arg2.
    In other word, the output of arg1 is the receiver of arg2.
    
    For example, to get city from user address:
    
        DOT(User.ADDRESS, Address.CITY).f(user); // return user.address.city
        
    `,
```

Comment (olhabn) [reply src/foam/mlang/mlang.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 25 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 24 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 23 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 22 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 21 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed. thanks

File: src/foam/mlang/expr/Ref.js (hunk 12 above)
Comment (jlhughes) [reply src/foam/mlang/expr/Ref.js]: The 'async' should be removed, correct? @kgrgreer

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 27]
```diff
@@ -81,61 +83,124 @@ foam.CLASS({
...
-        ClassInfo ci = of;
-        PropertyInfo p = (PropertyInfo) ci.getAxiomByName(propName[i]);
-        if ( i == propName.length - 1 ) {
-          return p;
-        }
+        PropertyInfo prop = (PropertyInfo) of.getAxiomByName(propName[i]);
+
+        if ( prop == null ) return null;
+
+        Method m = getFinderMethod(prop);
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: unused variable

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 28]
```diff
@@ -81,61 +83,124 @@ foam.CLASS({
...
+        expr = buildPropertyExpr(prop, expr);
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: please create new variable
```suggestion
        var propExpr = buildPropertyExpr(prop, expr);
```

File: src/foam/nanos/column/NestedPropertiesExpression.js [hunk 29]
```diff
@@ -81,61 +83,124 @@ foam.CLASS({
...
         }
+      `
+    },
+    {
+      name: 'getPropertyClassInfo',
+      javaType: 'ClassInfo',
+      javaThrows: [
+        'NoSuchMethodException',
+        'IllegalAccessException',
+        'InvocationTargetException'
```
Comment (lchanmann) [inline src/foam/nanos/column/NestedPropertiesExpression.js]: should be `javaThrows: [ 'java.lang.Exception' ]` because line 141 can return null.

Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 27 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed.

File: src/foam/nanos/column/NestedPropertiesExpression.js (hunk 28 above)
Comment (olhabn) [reply src/foam/nanos/column/NestedPropertiesExpression.js]: fixed.

**Review by lchanmann — APPROVED** [review]
Looks good. Thanks @olhabn

**Comment by olhabn** [comment]
I was fixing an exception. The cause of exception is that I didn't handle use case of retrieving something like "address.countryId", which looks totally legit to me, but we were using REF for countryId which caused the exception. so I added lastPropName check and if reference property is the last one in nested property's name string then it would be treated as not reference property.
please let me know if it makes sense or if I shouldn't allow such properties
I'd really appreciate a review with my last commit included @lchanmann 
thanks!

**Comment by lchanmann** [comment]
@olhabn you don't need that lastPropName argument and just return DOT in `if ( i === propName.length - 1)` block and move buildPropertyExpr call below it likes so
```
function (of, propName, i, expr) {
  var prop = of.getAxiomByName(propName[i]);
  if ( ! prop ) return null;

  if ( i === propName.length - 1 )
    return ! expr ? prop :
      foam.mlang.Expressions.create().DOT(expr, prop);

  var propExpr = this.buildPropertyExpr(prop, expr);
  return this.returnDotExprForNestedProperty(prop.of, propName, ++i, propExpr);
}
```

**Comment by olhabn** [comment]
> @olhabn you don't need that lastPropName argument and just return DOT in `if ( i === propName.length - 1)` block and move buildPropertyExpr call below it likes so
> 
> ```
> function (of, propName, i, expr) {
>   var prop = of.getAxiomByName(propName[i]);
>   if ( ! prop ) return null;
> 
>   if ( i === propName.length - 1 )
>     return ! expr ? prop :
>       foam.mlang.Expressions.create().DOT(expr, prop);
> 
>   var propExpr = this.buildPropertyExpr(prop, expr);
>   return this.returnDotExprForNestedProperty(prop.of, propName, ++i, propExpr);
> }
> ```

fixed. thanks!

**Review by lchanmann — APPROVED** [review]
@jlhughes @kgrgreer I'm good with this. Please take another look.

File: src/foam/mlang/expr/Ref.js (hunk 12 above)
Comment (kgrgreer) [reply src/foam/mlang/expr/Ref.js]: Don't make "async".

Comment (olhabn) [reply src/foam/mlang/expr/Ref.js]: fixed

**Review by jlhughes — APPROVED** [review]

**Pull request merged**